src/main/scala/chisel3/util/BitPat.scala

```diff
@@ -106,16 +122,30 @@ object BitPat {
   */
 sealed class BitPat(val value: BigInt, val mask: BigInt, width: Int) extends SourceInfoDoc {
   def getWidth: Int = width
+  def apply(x: Int): BitPat = macro SourceInfoTransform.xArg
+  def apply(x: Int, y: Int): BitPat = macro SourceInfoTransform.xyArg
   def === (that: UInt): Bool = macro SourceInfoTransform.thatArg
   def =/= (that: UInt): Bool = macro SourceInfoTransform.thatArg
-
+  def ## (that: BitPat): BitPat = macro SourceInfoTransform.thatArg
   override def equals(obj: Any): Boolean = {
     obj match {
       case y: BitPat => value == y.value && mask == y.mask && getWidth == y.getWidth
       case _ => false
     }
   }
 
+  /** @group SourceInfoTransformMacro */
+  def do_apply(x: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): BitPat = {
+    require(width > x && x >= 0, "index out of range")
+    BitPat(s"b${rawString(x)}")
+  }
+
+  /** @group SourceInfoTransformMacro */
+  def do_apply(x: Int, y: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): BitPat = {
+    require(width > y && y > x && x >= 0, "index out of range")
```

Comment:
This is backwards from how it works for `Bits`: https://scastie.scala-lang.org/NwxaWUqPRNSBOAxDJGYxoQ.

Also please provide a better error message like the above.

Reply:
I will rewrite this part.